### DISCUSSION TITLE
packagePatterns: ["*"] and excludePackagePatterns

### DISCUSSION THREAD

**Discussion by BasicWolf** [renovatebot/renovate#8330] [Request Help]
Hello, 

Could you please help understanding Renovate behaviour with the following `packageRules`:

```json
  "packageRules": [
    {
      "packagePatterns": ["*"],
      "excludePackagePatterns": ["org.junit.jupiter"],
      "managers": ["gradle-lite"],
      "groupName": "Gradle packages"
    }
  ]
```
Background story: we are using `palagePatterns: ["*"]` to group all Gradle packages updates in a single pull request. 
Once there was a need to exclude a single library from updates, and utilizing `excludePackagePatterns: ["library-name"]` seemed like a valid idea. 

The result however was, that Renovate created two separate pull requests:
1. With all packages updates excluding the one from `excludePackagePatterns`
2. With a single package update - the one from `excludePackagePatterns`

We have considered this a bug and filed it to Renovate (~~#8329~~, with a minimum reproduction [repository](https://github.com/BasicWolf/renovate-exclude-package-bug-2020_01_18)) however @viceice replied that it was not a bug and the following package rule should be used:

```json
"packageRules": [
  {
      "packagePatterns": ["org.junit.jupiter"],
      "managers": ["gradle-lite"],
      "enabled": false
    }
]
```

We will sure use the suggested rule, however doesn't the described behavior seems odd to you? The documentation to `excludePackagePatterns` states:
> Use this field if you want to have one or more package name patterns excluded in your package rule.

And we tried to do exactly what documentation says. Are we missing something here? Or is `"packagePatterns": ["*"]` a special case? Or...?

Thank you.

**Comment by viceice** [comment]
By default renovate will try to update all packages found in your repository.

After that renovate applies settings from matching packages rules, so it will group all your gradle packages excerpt the excluded packages.

`excludePackagePatterns` will not disable packges, so you need to add a disable rule.

So your final config should be something like:

```json
 "packageRules": [
    {
      "packagePatterns": ["*"],
      "managers": ["gradle-lite"],
      "groupName": "Gradle packages"
    },
    {
      "packagePatterns": ["org.junit.jupiter"],
      "managers": ["gradle-lite"],
      "enabled": false
    }
  ]
```

This will cause a group for all gradle packages (we discourage grouping major updates) and disable all packages which contains `org.junit.jupiter`.

Remember `packagePatterns` and `excludePackagePatterns`  are regex matchers, so the `.` dot matches every character. It also matches your string at any position, because you don't use `^` and `$` to mark start and end of string.

**Reply by rarkins** [reply]
Is it because the `gradle` package manager is also active?

**Reply by BasicWolf** [reply]
Thank you very much, @viceice, it's crystal clear now! 

@rarkins, doesn't look like `gradle` manager is active, we've migrated all our package rules to `managers: "gradle-lite"` and set "gradle-lite: { enabled: true }".

**Reply by rarkins** [reply]
Enabling gradle-lite doesn't disable gradle though. Have you explicitly disabled it anywhere?

**Reply by BasicWolf** [reply]
Thank you, @rarkins, ~~indeed we have not disabled the gradle manager!~~

Actually we had `enabledManagers: ` set with `gradle-lite`, but not `gradle`.
